compaction/compaction_manager.cc

```diff
@@ -1507,20 +1544,33 @@ future<> compaction_manager::perform_cleanup(owned_ranges_ptr sorted_owned_range
             t.schema()->ks_name(), t.schema()->cf_name()));
     }
 
+    if (sorted_owned_ranges->empty()) {
+        throw std::runtime_error("cleanup request failed: sorted_owned_ranges is empty");
+    }
+
+    // Called with compaction_disabled
     auto get_sstables = [this, &t, sorted_owned_ranges] () -> future<std::vector<sstables::shared_sstable>> {
         return seastar::async([this, &t, sorted_owned_ranges = std::move(sorted_owned_ranges)] {
-            auto schema = t.schema();
-            auto sstables = std::vector<sstables::shared_sstable>{};
-            const auto candidates = get_candidates(t);
-            std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(sstables), [&sorted_owned_ranges, schema] (const sstables::shared_sstable& sst) {
-                seastar::thread::maybe_yield();
-                return sorted_owned_ranges->empty() || needs_cleanup(sst, *sorted_owned_ranges, schema);
-            });
-            return sstables;
+            auto filter_candidates = [&] (const sstables::sstable_set& set) {
+                set.for_each_sstable([&] (const sstables::shared_sstable& sst) {
+                    update_sstable_cleanup_state(t, sst, *sorted_owned_ranges);
+                    seastar::thread::maybe_yield();
+                });
+            };
+            filter_candidates(t.main_sstable_set());
+            filter_candidates(t.maintenance_sstable_set());
+            // Some sstables may remain in sstables_requiring_cleanup
+            // for later processing if they can't be cleaned up right now.
+            // They are erased from sstables_requiring_cleanup by compacting.release_compacting
+            auto& cs = get_compaction_state(&t);
```

Comment:
By the way, sstable truncate will have to communicate with compaction manager if any being removed is needed for cleanup. Otherwise, sstables will remain alive until restart.

Reply:
good point

Reply:
added in v4